walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,78 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.stream.*;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<TimeRange> avaliableTimes = new ArrayList<>();
+        long duration = request.getDuration();
+        Collection<String> attendees = request.getAttendees();
+
+        // Edge Cases
+        if (duration >= TimeRange.WHOLE_DAY.duration()) {
+            return avaliableTimes;
+        }
+        else if (attendees.isEmpty() || events.isEmpty()){
+            avaliableTimes.add(TimeRange.WHOLE_DAY);
+            return avaliableTimes;
+        }
+
+        // Get Collection of unavaliableTimes times
+        Collection<TimeRange> unavaliableTimes = getUnavaliableTimes(events, attendees);
+        // Sort Unavaliable times by start time
+        List<TimeRange> unavTimesList = unavaliableTimes.stream().sorted(TimeRange.ORDER_BY_START).collect(Collectors.toList());
+        // Create Avaliable times
+        avaliableTimes = generateAvaliableTimes(unavTimesList, duration);
+        return avaliableTimes;
+    }
+
+    public Collection<TimeRange> getUnavaliableTimes(Collection<Event> events, Collection<String> attendees) {
+        Collection<TimeRange> badTimes = new ArrayList<>();
+        int i; // Used to ensure only one attendee has to attend another event to make the time unavaliable
+        for (Event event : events){
+            Set<String> eventAttendees = event.getAttendees();
```

Comment:
Could just replace this logic with checking if the two collections are disjoint? https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection)